xd.1
'\" t
.TH XD 1 "29 AUG 2000"
.UC 4
.SH NAME
xd \- extended dump and load utility
.SH SYNOPSIS
.B xd
[
.BI \-a addrfmt
.B \-c
.BI \-d label
.B \-l
.BI \-n datafmt
.B \-s
] [
.I infile
[
.I outfile
] ]
.SH DESCRIPTION
.B xd
dumps files in hexadecimal, decimal, or octal, optionally with ISO characters side by
side.
File addresses can likewise be displayed in hex, decimal, or octal notation.
.PP
.B xd
can
.I read
dump files in the same formats it writes and create binary files from
the data therein.  This allows you to dump a binary file with
.BR xd ,
edit it with your  favourite  text  editor,  then
make  a  new  binary  file containing whatever changes you've made.  When
creating a binary file,
.B xd
normally assumes you've only modified data in place (neither expanding
nor contracting the file) and verifies file addresses to guarantee
this.  However, a ``stream'' option is available which ignores file
addresses so you're free to insert and delete bytes at
will.
.B xd
thus turns your existing text editor into a binary file editor without
requiring you to learn any new commands.
.PP
Finally,
.B xd
can read a binary file and emit a C language data declaration which
contains the data from the file.  This is handy when you wish to embed
binary data within C programs.
.SH OPTIONS
.TP 10
.BI \-a addrfmt
Displays addresses in the dumped file as specified by
.IR addrfmt :
.B d
for decimal,
.B h
or
.B x
for hexadecimal, and
.B o
for octal.  By default, addresses are shown in hexadecimal.
.TP
.B \-c
Display ISO characters alongside the numeric dump.  Non-printing
characters (according to ISO 8859/1) are shown as periods.  The ISO
characters are separated from the dump by vertical bar
characters, which
.B xd
treats as comment delimiters when loading a file.
.TP
.BI \-d label
Dumps the file as a C language declaration of an array of
.B unsigned char
which, when compiled, will contain the same data as the binary input file.
The C array declaration is given the variable name
.IR label ,
or
.B xd_data
if no
.I label
specification is given.  Data are declared as individual bytes to
guarantee portability across architectures with different byte ordering
conventions, and are expressed as decimal numbers in lines of less than
80 characters for maximum portability among compilers.  The
C operator
.B sizeof
may be applied to the array
.I label
to obtain its length in bytes.
.TP
.B \-l
Load file from dump.
.B xd
reads a file in the same format it writes, creating a binary output
file.  Each line is assumed to begin with a file address terminated by
a colon.  Data bytes are separated by white space,
and any characters after a vertical bar
are ignored (thus discarding any ISO characters included in the
dump with the
.B \-c
option).  If the
.B \-s
option is not specified, each file address is checked against the
number of bytes written so far to guarantee that no bytes have been
added or deleted.
Files must be loaded with the
.B \-a
and
.B \-n
options set the same as when the file was dumped.
.TP
.BI \-n datafmt
Edits bytes in the dumped file as specified by
.IR datafmt :
.B d
for decimal,
.B h
or
.B x
for hexadecimal, and
.B o
for octal.  By default, bytes are shown in hexadecimal.
.TP
.B \-s
Stream input when loading file.  If this option is specified in conjunction
with the
.B \-l
option, file addresses are not verified when loading a file and, in fact,
need not be specified at all.  The input file is treated as a stream
of byte values separated by white space.  This option allows
you to create a dump with
.BR xd ,
edit it as you wish with any text editor, inserting and deleting bytes
in the file wherever you like, then create a binary file from the
modified dump with the command
.B xd \-l \-s
.IR "dumpfile outfile" .
.TP
.B \-u
Print how-to-call information.
.SH FILES
If no
.I infile
is specified,
.B xd
obtains its input from standard input; if no
.I outfile
is given, output is sent to standard output.  The input and
output are processed in a strictly serial manner regardless
of options; consequently
.B xd
may be used in pipelines without restrictions.
.SH BUGS
.PP
Input error checking in load mode might be improved.  Note that each
byte in load mode must be specified as a number in the same format
selected by the
.B \-n
option when the file was dumped,
separated by white space, and that the
.B \-a
and
.B \-n
options must be set the same when loading a file as
when it was dumped.
.SH "SEE ALSO"
.PD
.BR cc (1),
.BR od (1),
.BR iso_8859_1 (7)
.ne 10
.SH AUTHOR
.ce 2
John Walker
http://www.fourmilab.ch/
.PP
This software is in the public domain.
Permission to use, copy, modify, and distribute this software and its
documentation for any purpose and without fee is hereby granted,
without any conditions or restrictions.  This software is provided ``as
is'' without express or implied warranty.


sc/CHANGES
CHANGES BETWEEN 7.14 and 7.15:
Chuck Martin
	-Fixed "``" and "''" bug, introduced in 7.14 (they weren't being
	 accepted).  I guess 7.14 was released too soon after last minute
	 changes, without sufficient testing.
	-Fixed format bug that caused segfaults when there was no decimal
	 point in the format string, also introduced in 7.14 last minute
	 changes.
	-When entering insert mode with '+' or '-', and the cell already
	 contains a numeric value or expression, the edit is now saved in
	 the dot buffer, as it should be.
	-Don't save the current position (for the "`" and "'" commands)
	 during a goto in a macro.  Also, don't save the current position
	 during any movement command where the current cell doesn't change
	 (leave the saved position pointing to the last position before an
	 actual move).

CHANGES BETWEEN 7.13 and 7.14:
Chuck Martin
	-Added a braille option ("set braille" or ^Tb) with two different
	 modes, which makes use of braille displays with sc much easier
	 by judicious placement of the hardware cursor.
	-The "cellcur" option (current cell highlighting) is no longer
	 saved with a file, since it's more likely a user preference
	 and/or terminal dependent, and is more appropriate for use in a
	 user's .scrc.
	-Split the last fprintf() in list_frames() so that the outside
	 and inside ranges will be different (as they should be), even if
	 neither is a named range, since r_name() uses a single static
	 buffer.
	-The range copy command (/c) with only one argument will now use the
	 highlighted range as the source if one is highlighted.  Otherwise,
	 it will copy the last copied range, as before.
	-The range copy command (/c) with no arguments will now use the
	 highlighted range as the destination, if one is highlighted, and
	 the last copied range as the source.
	-The addnote command with only one argument will now use the
	 highlighted range if there is one, and the current cell otherwise,
	 as the range to link to the cell specified by the single argument.
	-The frame command with a single argument will use the highlighted
	 range, if any, as the inside range, and the single argument as
	 the outside range.  If no range is highlighted, the single argument
	 will be used as the inner range, and the existing outer range
	 containing the current cell will be used as the outer range, which
	 which was the previous behavior whenever a single range was speci-
	 fied.  If the current cell is not in an existing framed range, an
	 error message will be issued.
	-The frame command may now be used without any arguments, provided
	 a range is highlighted (which will be used as the inner range)
	 and the current cell is in a framed range whose outer range is to
	 be used as the default outer range.
	-Fixed a bug that prevented defining a new range with a name that
	 was the same as the beginning of a previously defined longer
	 range name.
	-A new command line option, -a, will skip the autorun macro if
	 one is defined.
	-Bug fix: if a cell contains a string expression, but no numeric
	 value, pressing "e" to edit the non-existent numeric value will
	 no longer bring up the string expression for editing (it wouldn't
	 work, anyway, because string expressions don't work with the "let"
	 command).
	-You can now use the cursor keys with ^Tr to set craction.
	-In numeric mode, entering insert mode by pressing a digit, '+',
	 or '-' will cause the cursor keys to terminate an entry and
	 move the cell cursor in the desired direction.  Switching to
	 edit mode will change the cursor keys back to their normal
	 function (moving the cursor in the command line or history), even
	 after switching back to insert mode.
	-Fixed @index function so that indexing by both row and column
	 will still work even if there is only one row or one column
	 (so that, for example, a macro could generically access the
	 cell in the first row and first column of a range without
	 regard for the actual dimensions of the range).
	-All flags are turned off in the clearent() function except
	 is_changed and is_cleared, which are turned on.  Also, the
	 is_cleared flag is turned off in the fill() function after the
	 cell is cleared and then assigned a new value.  This prevents
	 errors from appearing in expressions which reference cells in
	 the filled range whenever the sync_refs() function is called,
	 which happens any time any cell or range of cells is yanked,
	 deleted, pulled, copied, or moved.
	-Each instance of @ext now saves its own previous value, which
	 eliminates the longstanding bug that caused the last returned
	 value to be used for all instances if external functions were
	 enabled and later disabled.
	-Added a % command in edit mode that finds the matching parenthesis
	 of a pair, as in vi (only works for parentheses, not for braces
	 or brackets).
	-When starting a new command, the dot buffer will not be altered
	 unless the command starts out in edit mode (like the "e" and
	 "E" commands).  All other commands will only alter the dot buffer
	 after entering edit mode for the first time.  This allows things
	 like editing data, using "goto" to jump to a new cell, and then
	 "." to make the same change to the new cell (previously, the
	 "goto" command caused the previous edit to be lost from the dot
	 buffer).
	-Numeric arguments are now saved for use with the dot buffer when
	 editing a line, so the same argument will be used again when
	 repeating the action with the "." command.
	-If a command line is identical to the last command in the history,
	 it won't be stored in the history again.
	-The copy and move commands (/c and /m) no longer have an effect
	 on the delete buffer.  This allows you to do things like delete
	 a cell or range of cells, move things around in the spreadsheet,
	 and then pull the deleted cell(s) back in.
	-Implemented named buffers, "a" through "z" and "0" through "9",
	 selected with the " key.  Buffers "1" through "9" hold the last

xd.c
/*

		    Extended dump and load utility

			    by John Walker
		       http://www.fourmilab.ch/

		This program is in the public domain.

*/

#define Version \
                      "1.3  --  September 2000"


#include <stdio.h>
#include <ctype.h>
#include <string.h>
#ifdef _WIN32
#include <fcntl.h>
#include <io.h>
#endif

#define FALSE	0
#define TRUE	1

#define EOS     '\0'

static char addrformat[80] = "%6X";
static char scanaddr[80] = "%lx%c";

static char dataformat1[80] = "%02X";
static char scandata[80] = "%x%n%c";

static int bytesperline = 16, doublechar = FALSE,
	   dflen = 2, loading = FALSE, streaming = FALSE;
static unsigned long fileaddr;
static unsigned char lineecho[32];

/*  OUTLINE  --  Edit a line of binary data into the selected output
		 format.  */

static void outline(out, dat, len)
  FILE *out;
  unsigned char *dat;
  int len;
{
    char oline[132];
    int i;

    sprintf(oline, addrformat, fileaddr);
    strcat(oline, ":");
    for (i = 0; i < len; i++) {
	char outedit[80];

	sprintf(outedit, dataformat1, dat[i]);
        strcat(oline, (i == (bytesperline / 2)) ? "  " : " ");
	strcat(oline, outedit);
    }

    if (doublechar) {
	char oc[2];
	int shortfall = ((bytesperline - len) * (dflen + 1)) +
			(len <= (bytesperline / 2) ? 1 : 0);

	while (shortfall-- > 0) {
            strcat(oline, " ");
	}
	oc[1] = EOS;
        strcat(oline, " | ");
	for (i = 0; i < len; i++) {
	    int b = dat[i];

            /* Map non-printing characters to "." according to the
	       definitions for ISO 8859/1 Latin-1. */

            if ((b < ' ') || (b > '~' && b < 160)) {
                b = '.';
	    }

	    /* Many existing systems which support Latin-1 lack
	       a definition for character 160, the non-breaking
	       space.  Translate this to a space to avoid
	       confusion. */

	    if (b == 160) {
                b = ' ';
	    }
	    oc[0] = (char) b;
	    strcat(oline, oc);
	}
    }
    strcat(oline, "\n");
    fputs(oline, out);
}

/*  INTERPLINE	--  Interpret a line of input.	*/

static int interpline(line, lineno, out)
  char *line;
  int lineno;
  FILE *out;
{
    char *cp = line;
    char c;
    unsigned long lfaddr;
    int gfaddr = FALSE;

    /* Scan the line for a possible alternative format information
       following a vertical bar and delete it. */

    while ((c = *cp++) != EOS) {
        if (c == '|') {
	    cp[-1] = EOS;
	    break;
	}
    }

    /* Scan the line for a file address terminated by a colon.	Save
       the file address. */

    cp = line;

    while ((c = *cp++) != EOS) {
        if (c == ':') {
	    int sa;
	    char tchar;

	    cp[-1] = EOS;
	    sa = sscanf(line, scanaddr, &lfaddr, &tchar);
	    if (sa == 0 || (sa > 1 && tchar != EOS)) {
		fprintf(stderr,
                    "Bad file address \"%s\" on line %d:\n",
		    line, lineno);
		return FALSE;
	    }
	    gfaddr = TRUE;
            cp[-1] = ':';
	    break;
	}
    }
    if (!gfaddr) {
	cp = line;
    }
    if (!streaming) {
	if (!gfaddr) {
            fprintf(stderr, "File address missing on line %d:\n", lineno);
            fprintf(stderr, "%s\n", line);
	    return FALSE;
	}
	if (lfaddr != fileaddr) {
            fprintf(stderr, "File address sequence error on line %d.\n",
		lineno);
            fprintf(stderr, "  Expected ");
	    fprintf(stderr, addrformat, fileaddr);
            fprintf(stderr, ", received ");
	    fprintf(stderr, addrformat, lfaddr);
            fprintf(stderr, ".\n");
            fprintf(stderr, "%s\n", line);
	    return FALSE;
	}
    }

    while ((c = *cp++) != EOS) {
	if (!isspace(c)) {
	    int scanl, nscan, dvalue;
	    char termchar;

	    if (((scanl = sscanf(cp - 1, scandata, &dvalue, &nscan, &termchar)) == 0) ||
		(dvalue < 0) || (dvalue > 255) || 
		(scanl == 2 && !isspace(termchar))) {
                fprintf(stderr, "Improper value, \"%s\" on line %d:\n",
		    cp - 1, lineno);
                fprintf(stderr, "%s\n", line);
                fprintf(stderr, "Bytes must be specified as digits separated by white space.\n");
		return FALSE;
	    }
	    putc((char) dvalue, out);
	    fileaddr++;
	    cp += nscan;
	}
    }
    return TRUE;
}

/*  Main program  */

int main(argc, argv)
  int argc; char *argv[];
{
    int i, b, bp, cdata = FALSE, f = 0;
    char *cp, *clabel, opt;
    FILE *in = stdin, *out = stdout;

    for (i = 1; i < argc; i++) {
	cp = argv[i];
        if (*cp == '-') {
	    opt = *(++cp);
	    if (islower(opt)) {
		opt = (char) toupper(opt);
	    }
	    switch (opt) {

                case 'A':             /* -Af  --  Set address format */
		    opt = cp[1];
		    if (islower(opt)) {
			opt = (char) toupper(opt);
		    }
		    switch (opt) {
                        case 'D':
                            strcpy(addrformat, "%8d");
                            strcpy(scanaddr, "%ld%c");
			    break;

                        case 'H':
                        case 'X':
                            strcpy(addrformat, "%6X");
                            strcpy(scanaddr, "%lx%c");
			    break;

                        case 'O':
                            strcpy(addrformat, "%8o");
                            strcpy(scanaddr, "%lo%c");
			    break;

			default:
			    fprintf(stderr,
                            "Invalid address format '%c'.  Must be D, H, or O.\n", cp[1]);
			    return 2;
		    }
		    break;

                case 'C':
		    doublechar = TRUE;
		    break;

                case 'D':
		    cdata = TRUE;
		    clabel = cp + 1;
		    break;

                case 'L':
		    loading = TRUE;
		    break;

                case 'N':             /* -Nf  --  Set numeric dump format */
		    opt = cp[1];
		    if (islower(opt)) {
			opt = (char) toupper(opt);
		    }
		    switch (opt) {
                        case 'D':
                            strcpy(dataformat1, "%3d");
                            strcpy(scandata, "%d%n%c");
			    break;

                        case 'H':
                        case 'X':
                            strcpy(dataformat1, "%02X");
                            strcpy(scandata, "%x%n%c");
			    break;

                        case 'O':
                            strcpy(dataformat1, "%03o");
                            strcpy(scandata, "%o%n%c");
			    break;

			default:
			    fprintf(stderr,
                            "Invalid numeric dump format '%c'.  Must be D, H, or O.\n", cp[1]);
			    return 2;
		    }
		    break;

                case 'S':
		    streaming = TRUE;
		    break;

                case '?':
                case 'H':
                case 'U':
    fprintf(stderr, "XD  --  Extended dump.  Call with xd [input [output]]\n");
    fprintf(stderr, "\n");
    fprintf(stderr, "        Options:\n");
    fprintf(stderr, "             -af     Print addresses in f = Decimal, Hex, or Octal\n");
    fprintf(stderr, "             -c      Dump as ISO characters\n");
    fprintf(stderr, "             -dlabel Dump as a C data declaration\n");
    fprintf(stderr, "             -l      Load file from hex dump\n");
    fprintf(stderr, "             -nf     Numeric dump in f = Decimal, Hex, or Octal\n");
    fprintf(stderr, "             -s      Stream load (don't check file addresses)\n");
    fprintf(stderr, "             -u      Print this message\n");
    fprintf(stderr, "\nBy John Walker, http://www.fourmilab.ch/\n");
    fprintf(stderr,"Version %s\n", Version);
		    return 0;
	    }
	} else {
	    switch (f) {
		case 0:

		/** Warning!  On systems which distinguish text mode and
		    binary I/O (MS-DOS, Macintosh, etc.) the modes in these
		    open statements will have to be made conditional based
		    upon whether an encode or decode is being done, which
                    will have to be specified earlier.  But it's worse: if
		    input or output is from standard input or output, the 
		    mode will have to be changed on the fly, which is
                    generally system and compiler dependent.  'Twasn't me
                    who couldn't conform to Unix CR/LF convention, so 
                    don't ask me to write the code to work around
                    Apple and Microsoft's incompatible standards.

		    This file contains code, conditional on _WIN32, which
		    sets binary mode using the method prescribed by
                    Microsoft Visual C 1.52 ("Monkey C"); this may
                    require modification if you're using a different
		    compiler or release of Monkey C.  */

                    if ((in = fopen(cp, loading ? "r" : "rb")) == NULL) {
                        fprintf(stderr, "Cannot open input file %s\n", cp);
			return 2;
		    }
		    f++;
		    break;

		case 1:
                    if ((out = fopen(cp, loading ? "wb" : "w")) == NULL) {
                        fprintf(stderr, "Cannot open output file %s\n", cp);
			return 2;
		    }
		    f++;
		    break;

		default:
                    fprintf(stderr, "Too many file names specified.\n");
	    }
	}
    }

#ifdef _WIN32

    /*  If input is from standard input and we aren't loading
	from a dump file, set the input file mode to binary.  */

    if ((in == stdin) && (!loading)) {
	_setmode(_fileno(in), _O_BINARY);
    }

    /*  If output is to standard output and we're loading a
	binary file from a dump, set the output file mode to
	binary.  */

    if ((out == stdout) && (loading)) {
	_setmode(_fileno(out), _O_BINARY);
    }
#endif

    bp = 0;
    fileaddr = 0;

    if (loading) {
	char in_line[256];
	int lineno = 0;

	while (fgets(in_line, (sizeof in_line) - 2, in)) {
	    lineno++;
	    if (!interpline(in_line, lineno, out)) {
		fclose(out);
		return 2;
	    }
	}
    } else {
	if (cdata) {
	    char cout[80];

            fprintf(out, "unsigned char %s[] = {\n",
                clabel[0] == EOS ? "xd_data" : clabel);
            strcpy(cout, "    ");

	    while ((b = getc(in)) != EOF) {
		if (strlen(cout) > 72) {
                    fprintf(out, "%s\n", cout);
                    strcpy(cout, "    ");
		}
                sprintf(cout + strlen(cout), "%d,", b);
	    }
	    if (strlen(cout) > 4) {
		cout[strlen(cout) - 1] = EOS; /* Strip trailing comma */
                fprintf(out, "%s\n", cout);
	    }
            fprintf(out, "};\n");
	} else {
	    while ((b = getc(in)) != EOF) {
		if (bp >= bytesperline) {
		    outline(out, lineecho, bp);
		    bp = 0;
		    fileaddr += bytesperline;
		}
		lineecho[bp++] = (char) b;
	    }

	    if (bp > 0) {
		outline(out, lineecho, bp);
	    }
	}
    }
    return 0;
}
